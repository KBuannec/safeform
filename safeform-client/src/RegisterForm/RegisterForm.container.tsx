import RegisterForm from "./RegisterForm";
import { useRegisterForm } from "./RegisterForm.hook";

const RegisterFormContainer = () => {
    const { form, errors, success, handleChange, handleSubmit } = useRegisterForm();

    return (
        <div className="register-form-wrapper">
            <RegisterForm form={form} errors={errors} success={success} onChange={handleChange} onSubmit={handleSubmit} />
        </div>
    );
};

export default RegisterFormContainer;