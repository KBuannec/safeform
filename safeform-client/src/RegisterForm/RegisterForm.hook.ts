import { useState } from 'react';
import { z } from 'zod';

const registerSchema = z.object({
  name: z.string().min(1, 'Nom requis'),
  email: z.string().email('Email invalide'),
  password: z.string().min(6, 'Mot de passe trop court'),
});

export type RegisterFormData = z.infer<typeof registerSchema>;

export const useRegisterForm = () => {
    const [form, setForm] = useState<RegisterFormData>({
        name: '',
        email: '',
        password: '',
    });

    const [errors, setErrors] = useState<Partial<Record<keyof RegisterFormData, string>>>({});
    const [success, setSuccess] = useState<string | null>(null);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setForm({ ...form, [e.target.name]: e.target.value });
    };

    const handleSubmit = async (e: React.FormEvent) => {
        e.preventDefault();
        const result = registerSchema.safeParse(form);

        if (!result.success) {
            const fieldsErrors = result.error.flatten().fieldErrors;
            setErrors({
                name: fieldsErrors.name?.[0],
                email: fieldsErrors.email?.[0],
                password: fieldsErrors.password?.[0],
            });
            return;
        }

        setErrors({});
        try {
            const res = await fetch('http://localhost:3000/register', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                credentials: 'include',
                body: JSON.stringify(form),
            });

            const data = await res.json();

            if (!res.ok) {
                setSuccess(null);
                setErrors({ email: data.message || 'Erreur inconnue' });
                return;
            }

            setSuccess('Compte créé avec succès !');
            setForm({ name: '', email: '', password: '' })
        } catch (err) {
            console.error(err);
            setSuccess("Une erreur est survenue");
        }
    };

    return {
        form,
        errors,
        success,
        handleChange,
        handleSubmit,
    };
};