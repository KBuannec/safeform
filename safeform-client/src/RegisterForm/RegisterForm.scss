.register-form {
    background-color: #fff;
    padding: 2rem;
    border-radius: 0.75rem;
    box-shadow: 0 2px 12px rgba(0, 0, 0, 0.05);
    max-width: 500px;
    margin: 2rem auto;
    display: flex;
    flex-direction: column;
    gap: 1rem;

    h2 {
        text-align: center;
        color: #1e293b;
        margin-bottom: 1rem;
    }

    label {
        font-weight: 600;
        color: #334155;
    }

    input {
        padding: 0.75rem;
        border: 1px solid #cbd5e1;
        border-radius: 0.375rem;
        font-size: 1rem;
        transition: border-color 0.2s, box-shadow 0.2s;

        &:focus {
            border-color: #2563eb;
            box-shadow: 0 0 0 2px rgba(37, 99, 235, 0.2);
            outline: none;
        }
    }

    button {
        padding: 0.75rem;
        background-color: #2563eb;
        color: white;
        border: none;
        border-radius: 0.375rem;
        font-weight: 600;
        cursor: pointer;
        transition: background-color 0.2s, transform 0.1s ease;

        &:hover {
            background-color: #1e40af;
        }

        &:active {
            transform: scale(0.98);
        }
    }

    .error {
        color: #dc2626;
        font-size: 0.875rem;
    }

    .success {
        color: #16a34a;
        font-size: 0.875rem;
        text-align: center;
        animation: fadeIn 0.6s ease-in-out;
    }

    @keyframes fadeIn {
        from {
            opacity: 0;
            transform: translateY(-5px);
        }

        to {
            opacity: 1;
            transform: translateY(0);
        }
    }
}

.input-wrapper {
  position: relative;
  width: 100%;

  input {
    width: 100%;
    padding: 0.6rem 2.5rem 0.6rem 0.6rem;
    border: 1px solid #ccc;
    border-radius: 6px;
    font-size: 1rem;
    box-sizing: border-box;
  }

  .eye-icon {
    position: absolute;
    top: 50%;
    right: 0.75rem;
    transform: translateY(-50%);
    cursor: pointer;
    font-size: 1.1rem;
    user-select: none;
    color: #666;

    &:hover {
      color: #111;
    }
  }
}
