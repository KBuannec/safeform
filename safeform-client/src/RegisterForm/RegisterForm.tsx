import { useState } from 'react';
import './RegisterForm.scss';

interface RegisterFormProps {
    form: {
        name: string;
        email: string;
        password: string;
    };
    errors: {
        name?: string;
        email?: string;
        password?: string;
    };
    success: string | null;
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
    onSubmit: React.FormEventHandler<HTMLFormElement>;
}

const RegisterForm = ({ form, errors, success, onChange, onSubmit }: RegisterFormProps) => {

    const [showPassword, setShowPassword] = useState(false);

    const togglePasswordVisibility = () => {
        setShowPassword(prev => !prev);
    };

    return (
        <form className='register-form' onSubmit={onSubmit} role='form'>
            <h2>Créer un compte</h2>

            <label htmlFor='name'>Nom</label>
            <input id='name' name='name' value={form.name} onChange={onChange} placeholder='Nom' />
            {errors.name && <p className="error">{errors.name}</p>}

            <label htmlFor='email'>Email</label>
            <input id='email' name='email' value={form.email} onChange={onChange} placeholder='test@example.fr'/>
            {errors.email && <p className="error">{errors.email}</p>}

            <label htmlFor='password'>Mot de passe</label>
            <div className="input-wrapper">
                <input
                    id='password'
                    name='password'
                    type={showPassword ? 'text' : 'password'}
                    value={form.password}
                    onChange={onChange}
                    placeholder='••••••••'
                />
                <span className='eye-icon' onClick={togglePasswordVisibility}>
                    {showPassword ? '🙈' : '👁️'}
                </span>
            </div>
            {errors.password && <p className="error">{errors.password}</p>}

            <button type='submit'>S'inscrire</button>

            {success && <p className='success'>{success}</p>}
        </form>
    );
};

export default RegisterForm;