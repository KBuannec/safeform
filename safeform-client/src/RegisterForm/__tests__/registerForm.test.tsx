import { describe, it, expect, vi } from 'vitest'
import { render, screen, fireEvent } from '@testing-library/react'
import RegisterForm from '../RegisterForm'

describe('RegisterForm component', () => {
    const baseProps = {
        form: {
            name: '',
            email: '',
            password: '',
        },
        errors: {},
        success: null,
        onChange: vi.fn(),
        onSubmit: vi.fn(),
    }

    it('affiche les champs et le titre', () => {
        render(<RegisterForm {...baseProps} />)

        expect(screen.getByRole('heading', { name: /créer un compte/i })).toBeInTheDocument()
        expect(screen.getByLabelText(/nom/i)).toBeInTheDocument()
        expect(screen.getByLabelText(/email/i)).toBeInTheDocument()
        expect(screen.getByLabelText(/mot de passe/i)).toBeInTheDocument()
    })

    it("affiche les erreurs si présentes", () => {
        render(
            <RegisterForm
                {...baseProps}
                errors={{
                    name: 'Nom requis',
                    email: 'Email invalide',
                    password: 'Mot de passe trop court',
                }}
            />
        )

        expect(screen.getByText('Nom requis')).toBeInTheDocument()
        expect(screen.getByText('Email invalide')).toBeInTheDocument()
        expect(screen.getByText('Mot de passe trop court')).toBeInTheDocument()
    })

    it("permet d'afficher le mot de passe", () => {
        render(<RegisterForm {...baseProps} />)

        const eyeIcon = screen.getByText('👁️')
        fireEvent.click(eyeIcon)

        expect(screen.getByPlaceholderText('••••••••')).toHaveAttribute('type', 'text')
    })

    it('soumet le formulaire', () => {
        render(<RegisterForm {...baseProps} />)

        fireEvent.submit(screen.getByRole('form'))
        expect(baseProps.onSubmit).toHaveBeenCalled()
    })

    it('correspond au snapshot', () => {
        const { asFragment } = render(<RegisterForm {...baseProps} />)
        expect(asFragment()).toMatchSnapshot()
    })
})