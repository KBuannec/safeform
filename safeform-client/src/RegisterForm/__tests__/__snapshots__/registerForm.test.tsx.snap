// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterForm component > correspond au snapshot 1`] = `
<DocumentFragment>
  <form
    class="register-form"
    role="form"
  >
    <h2>
      Créer un compte
    </h2>
    <label
      for="name"
    >
      Nom
    </label>
    <input
      id="name"
      name="name"
      placeholder="Nom"
      value=""
    />
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      placeholder="test@example.fr"
      value=""
    />
    <label
      for="password"
    >
      Mot de passe
    </label>
    <div
      class="input-wrapper"
    >
      <input
        id="password"
        name="password"
        placeholder="••••••••"
        type="password"
        value=""
      />
      <span
        class="eye-icon"
      >
        👁️
      </span>
    </div>
    <button
      type="submit"
    >
      S'inscrire
    </button>
  </form>
</DocumentFragment>
`;
